spec/ruby-push-notifications/wns/wns_pusher_spec.rb

```diff
@@ -0,0 +1,58 @@
+module RubyPushNotifications
+  module WNS
+    describe WNSPusher do
+      let(:access_token) { 'token' }
+      let(:pusher) { WNSPusher.new(access_token) }
+
+      describe '#push' do
+        let(:device_urls) { [generate(:wns_device_url)] }
+        let(:notifications) { build_list(:wns_notification, 2, device_urls: ['http://hk2.notify.windows.com/1'], data: { message: { value1: 'hello' } }).to_a }
+        let(:raw_notification) { build :wns_notification }
+        let(:toast_data) { { title: 'Title', message: 'Hello WNS World!', type: :toast } }
+        let(:toast_notification) { build :wns_notification, device_urls: device_urls, data: toast_data }
+        let(:tile_data) { { message: 'Hello WNS World!', image: 'http://image.com/image.jpg', type: :tile } }
+        let(:tile_notification) { build :wns_notification, device_urls: device_urls, data: tile_data }
+        let(:badge_data) { { value: 10, type: :badge } }
+        let(:badge_notification) { build :wns_notification, device_urls: device_urls, data: badge_data }
+
+        let(:headers) {
+          {
+            'x-wns-notificationstatus' => 'Received',
+            'x-wns-deviceconnectionstatus' => 'Connected',
+            'x-wns-status' => 'Received'
+          }
+        }
+        let(:wns_connection) { RubyPushNotifications::WNS::WNSConnection }
+        let(:response) { [ { device_url: 'http://hk2.notify.windows.com/1', headers: headers, code: 200 } ]}
+
+        before do
+          stub_request(:post, %r{hk2.notify.windows.com}).
+            to_return status: [200, 'OK'], headers: headers
+        end
+
+        it 'submits every notification' do
+          pusher.push [raw_notification, toast_notification, tile_notification, badge_notification]
+          wns_connection::WP_TARGETS.keys.each do |push_type|
+            headers = {
+              wns_connection::CONTENT_TYPE_HEADER => wns_connection::CONTENT_TYPE[push_type],
+              wns_connection::X_WNS_TYPE_HEADER => wns_connection::WP_TARGETS[push_type],
+              wns_connection::CONTENT_LENGTH_HEADER => public_send("#{push_type}_notification").as_wns_xml.length.to_s,
```

Comment:
Where is this `public_send` defined? I can't find it

Reply:
It's a ruby method: https://apidock.com/ruby/Object/public_send